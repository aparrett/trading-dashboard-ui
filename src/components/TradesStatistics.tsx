import { Box, Grid, Paper } from '@material-ui/core'
import { FC } from 'react'
import { TradeWithPL } from '../types'
import { getProfitClass } from '../util/getProfitClass'
import { roundTwo } from '../util/roundPenny'

interface TradesStatisticsProps {
    trades: TradeWithPL[]
}

export const TradesStatistics: FC<TradesStatisticsProps> = ({ trades }) => {
    const profit = trades.reduce((acc, trade) => acc + trade.profitLoss, 0)
    const tradesWithoutNeutrals = trades.filter((trade) => trade.profitLoss > 10 || trade.profitLoss < -10)
    const winners = tradesWithoutNeutrals.filter((trade) => trade.profitLoss >= 0)
    const losers = tradesWithoutNeutrals.filter((trade) => trade.profitLoss < 0)
    const averageWinner =
        winners.length === 0 ? 0 : winners.reduce((acc, trade) => acc + trade.profitLoss, 0) / winners.length
    const averageLoser =
        losers.length === 0 ? 0 : losers.reduce((acc, trade) => acc + trade.profitLoss, 0) / losers.length

    const winPercentage =
        winners.length > 0 && tradesWithoutNeutrals.length > 0
            ? Math.round((winners.length / tradesWithoutNeutrals.length) * 100)
            : 0
    return (
        <Grid item sm={12} md={6}>
            <Paper className="paper">
                <Grid container>
                    <Grid item xs={4}>
                        <Box display="flex" justifyContent="space-between">
                            <div>Trades:</div>
                            <div data-testid="tradesCount">{trades.length}</div>
                        </Box>
                        <Box display="flex" justifyContent="space-between">
                            <div>Winning Trades:</div>
                            <div data-testid="winningTradesCount">{winners.length}</div>
                        </Box>
                        <Box display="flex" justifyContent="space-between">
                            <div>Losing Trades:</div>
                            <div data-testid="losingTradesCount">{losers.length}</div>
                        </Box>
                        <Box display="flex" justifyContent="space-between">
                            <div>Neutral Trades:</div>
                            <div data-testid="neutralTradesCount">{trades.length - tradesWithoutNeutrals.length}</div>
                        </Box>
                    </Grid>
                    <Grid item xs={2} />
                    <Grid item xs={4}>
                        <Box display="flex" justifyContent="space-between">
                            <div>Profit:</div>
                            <div data-testid="profit" className={getProfitClass(profit)}>{roundTwo(profit)}</div>
                        </Box>
                        <Box display="flex" justifyContent="space-between">
                            <div>Average Winner:</div>
                            <div data-testid="avgWinner" className={getProfitClass(averageWinner)}>{roundTwo(averageWinner)}</div>
                        </Box>
                        <Box display="flex" justifyContent="space-between">
                            <div>Average Loser:</div>
                            <div data-testid="avgLoser" className={getProfitClass(averageLoser)}>{roundTwo(averageLoser)}</div>
                        </Box>
                        <Box display="flex" justifyContent="space-between">
                            <div>Win Percentage:</div>
                            <div data-testid="winPerc">{winPercentage}%</div>
                        </Box>
                    </Grid>
                    <Grid item xs={2} />
                </Grid>
            </Paper>
        </Grid>
    )
}
