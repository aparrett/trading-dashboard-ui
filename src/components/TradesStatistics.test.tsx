import { render, screen } from '@testing-library/react'
import { TradesStatistics } from './TradesStatistics'

// This component doesn't care about close date, so it is left out. If we wanted to add it, we'd have to change it with profit loss every time or it could be confusing.
const trade = {
    entry: 1,
    id: 1,
    openDate: expect.any(String),
    profitLoss: 1,
    quantity: 1,
    side: 'Long',
    symbol: 'TEST',
    traderId: 1
}

it('renders correctly', () => {
    const wrapper = render(<TradesStatistics trades={[]} />)
    expect(wrapper).toMatchSnapshot()
})

it('renders correctly with a trade', () => {
    const wrapper = render(<TradesStatistics trades={[trade]} />)
    expect(wrapper).toMatchSnapshot()
})

it('should calculate neutrals correctly - 0 trades', () => {
    render(<TradesStatistics trades={[]} />)
    const el = screen.getByTestId('neutralTradesCount')
    expect(el.textContent).toEqual('0')
})

it('should calculate neutrals correctly - two neutrals and one win', () => {
    const trade1 = { ...trade, profitLoss: 100 }
    render(<TradesStatistics trades={[trade1, trade, trade]} />)
    const el = screen.getByTestId('neutralTradesCount')
    expect(el.textContent).toEqual('2')
})

it('should calculate average winner correctly - basic', () => {
    const trade1 = { ...trade, profitLoss: 100 }
    render(<TradesStatistics trades={[trade1, trade1]} />)
    const el = screen.getByTestId('avgWinner')
    expect(el.textContent).toEqual('100')
})

it('should calculate average winner correctly - 0 winners', () => {
    const trade1 = { ...trade, profitLoss: -100 }
    render(<TradesStatistics trades={[trade1]} />)
    const el = screen.getByTestId('avgWinner')
    expect(el.textContent).toEqual('0')
})

it('should calculate average winner correctly - ignores neutrals', () => {
    const trade1 = { ...trade, profitLoss: 1 }
    const trade2 = { ...trade, profitLoss: 100 }
    render(<TradesStatistics trades={[trade1, trade2]} />)
    const el = screen.getByTestId('avgWinner')
    expect(el.textContent).toEqual('100')
})

it('should calculate average winner correctly - ignores losers', () => {
    const trade1 = { ...trade, profitLoss: -100 }
    const trade2 = { ...trade, profitLoss: 100 }
    render(<TradesStatistics trades={[trade1, trade2]} />)
    const el = screen.getByTestId('avgWinner')
    expect(el.textContent).toEqual('100')
})

it('should calculate average loser correctly - basic', () => {
    const trade1 = { ...trade, profitLoss: -100 }
    render(<TradesStatistics trades={[trade1, trade1]} />)
    const el = screen.getByTestId('avgLoser')
    expect(el.textContent).toEqual('-100')
})

it('should calculate average loser correctly - 0 losers', () => {
    const trade1 = { ...trade, profitLoss: 100 }
    render(<TradesStatistics trades={[trade1]} />)
    const el = screen.getByTestId('avgLoser')
    expect(el.textContent).toEqual('0')
})

it('should calculate average loser correctly - ignores neutrals', () => {
    const trade1 = { ...trade, profitLoss: 1 }
    const trade2 = { ...trade, profitLoss: -100 }
    render(<TradesStatistics trades={[trade1, trade2]} />)
    const el = screen.getByTestId('avgLoser')
    expect(el.textContent).toEqual('-100')
})

it('should calculate average loser correctly - ignores winners', () => {
    const trade1 = { ...trade, profitLoss: -100 }
    const trade2 = { ...trade, profitLoss: 100 }
    render(<TradesStatistics trades={[trade1, trade2]} />)
    const el = screen.getByTestId('avgLoser')
    expect(el.textContent).toEqual('-100')
})

it('should calculate win perc correctly - 100%', () => {
    const trade1 = { ...trade, profitLoss: 100 }
    render(<TradesStatistics trades={[trade1]} />)
    const el = screen.getByTestId('winPerc')
    expect(el.textContent).toEqual('100%')
})

it('should calculate win perc correctly - 0%', () => {
    const trade1 = { ...trade, profitLoss: 0 }
    render(<TradesStatistics trades={[trade1]} />)
    const el = screen.getByTestId('winPerc')
    expect(el.textContent).toEqual('0%')
})

it('should calculate win perc correctly - 50%', () => {
    const trade1 = { ...trade, profitLoss: -100 }
    const trade2 = { ...trade, profitLoss: 100 }
    render(<TradesStatistics trades={[trade1, trade2]} />)
    const el = screen.getByTestId('winPerc')
    expect(el.textContent).toEqual('50%')
})

it('should calculate win perc correctly - ignores neutrals', () => {
    const trade1 = { ...trade, profitLoss: -100 }
    const trade2 = { ...trade, profitLoss: 100 }
    const trade3 = { ...trade, profitLoss: 0 }
    render(<TradesStatistics trades={[trade1, trade2, trade3]} />)
    const el = screen.getByTestId('winPerc')
    expect(el.textContent).toEqual('50%')
})
