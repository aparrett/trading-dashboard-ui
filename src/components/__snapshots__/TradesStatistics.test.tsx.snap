// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-sm-12 MuiGrid-grid-md-6"
      >
        <div
          class="MuiPaper-root paper MuiPaper-elevation1 MuiPaper-rounded"
        >
          <div
            class="MuiGrid-root MuiGrid-container"
          >
            <div
              class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-4"
            >
              <div
                class="MuiBox-root MuiBox-root-1"
              >
                <div>
                  Trades:
                </div>
                <div
                  data-testid="tradesCount"
                >
                  0
                </div>
              </div>
              <div
                class="MuiBox-root MuiBox-root-2"
              >
                <div>
                  Winning Trades:
                </div>
                <div
                  data-testid="winningTradesCount"
                >
                  0
                </div>
              </div>
              <div
                class="MuiBox-root MuiBox-root-3"
              >
                <div>
                  Losing Trades:
                </div>
                <div
                  data-testid="losingTradesCount"
                >
                  0
                </div>
              </div>
              <div
                class="MuiBox-root MuiBox-root-4"
              >
                <div>
                  Neutral Trades:
                </div>
                <div
                  data-testid="neutralTradesCount"
                >
                  0
                </div>
              </div>
            </div>
            <div
              class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-2"
            />
            <div
              class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-4"
            >
              <div
                class="MuiBox-root MuiBox-root-5"
              >
                <div>
                  Profit:
                </div>
                <div
                  class=""
                  data-testid="profit"
                >
                  0
                </div>
              </div>
              <div
                class="MuiBox-root MuiBox-root-6"
              >
                <div>
                  Average Winner:
                </div>
                <div
                  class=""
                  data-testid="avgWinner"
                >
                  0
                </div>
              </div>
              <div
                class="MuiBox-root MuiBox-root-7"
              >
                <div>
                  Average Loser:
                </div>
                <div
                  class=""
                  data-testid="avgLoser"
                >
                  0
                </div>
              </div>
              <div
                class="MuiBox-root MuiBox-root-8"
              >
                <div>
                  Win Percentage:
                </div>
                <div
                  data-testid="winPerc"
                >
                  0
                  %
                </div>
              </div>
            </div>
            <div
              class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-2"
            />
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="MuiGrid-root MuiGrid-item MuiGrid-grid-sm-12 MuiGrid-grid-md-6"
    >
      <div
        class="MuiPaper-root paper MuiPaper-elevation1 MuiPaper-rounded"
      >
        <div
          class="MuiGrid-root MuiGrid-container"
        >
          <div
            class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-4"
          >
            <div
              class="MuiBox-root MuiBox-root-1"
            >
              <div>
                Trades:
              </div>
              <div
                data-testid="tradesCount"
              >
                0
              </div>
            </div>
            <div
              class="MuiBox-root MuiBox-root-2"
            >
              <div>
                Winning Trades:
              </div>
              <div
                data-testid="winningTradesCount"
              >
                0
              </div>
            </div>
            <div
              class="MuiBox-root MuiBox-root-3"
            >
              <div>
                Losing Trades:
              </div>
              <div
                data-testid="losingTradesCount"
              >
                0
              </div>
            </div>
            <div
              class="MuiBox-root MuiBox-root-4"
            >
              <div>
                Neutral Trades:
              </div>
              <div
                data-testid="neutralTradesCount"
              >
                0
              </div>
            </div>
          </div>
          <div
            class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-2"
          />
          <div
            class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-4"
          >
            <div
              class="MuiBox-root MuiBox-root-5"
            >
              <div>
                Profit:
              </div>
              <div
                class=""
                data-testid="profit"
              >
                0
              </div>
            </div>
            <div
              class="MuiBox-root MuiBox-root-6"
            >
              <div>
                Average Winner:
              </div>
              <div
                class=""
                data-testid="avgWinner"
              >
                0
              </div>
            </div>
            <div
              class="MuiBox-root MuiBox-root-7"
            >
              <div>
                Average Loser:
              </div>
              <div
                class=""
                data-testid="avgLoser"
              >
                0
              </div>
            </div>
            <div
              class="MuiBox-root MuiBox-root-8"
            >
              <div>
                Win Percentage:
              </div>
              <div
                data-testid="winPerc"
              >
                0
                %
              </div>
            </div>
          </div>
          <div
            class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-2"
          />
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`renders correctly with a trade 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-sm-12 MuiGrid-grid-md-6"
      >
        <div
          class="MuiPaper-root paper MuiPaper-elevation1 MuiPaper-rounded"
        >
          <div
            class="MuiGrid-root MuiGrid-container"
          >
            <div
              class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-4"
            >
              <div
                class="MuiBox-root MuiBox-root-9"
              >
                <div>
                  Trades:
                </div>
                <div
                  data-testid="tradesCount"
                >
                  1
                </div>
              </div>
              <div
                class="MuiBox-root MuiBox-root-10"
              >
                <div>
                  Winning Trades:
                </div>
                <div
                  data-testid="winningTradesCount"
                >
                  0
                </div>
              </div>
              <div
                class="MuiBox-root MuiBox-root-11"
              >
                <div>
                  Losing Trades:
                </div>
                <div
                  data-testid="losingTradesCount"
                >
                  0
                </div>
              </div>
              <div
                class="MuiBox-root MuiBox-root-12"
              >
                <div>
                  Neutral Trades:
                </div>
                <div
                  data-testid="neutralTradesCount"
                >
                  1
                </div>
              </div>
            </div>
            <div
              class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-2"
            />
            <div
              class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-4"
            >
              <div
                class="MuiBox-root MuiBox-root-13"
              >
                <div>
                  Profit:
                </div>
                <div
                  class="green"
                  data-testid="profit"
                >
                  1
                </div>
              </div>
              <div
                class="MuiBox-root MuiBox-root-14"
              >
                <div>
                  Average Winner:
                </div>
                <div
                  class=""
                  data-testid="avgWinner"
                >
                  0
                </div>
              </div>
              <div
                class="MuiBox-root MuiBox-root-15"
              >
                <div>
                  Average Loser:
                </div>
                <div
                  class=""
                  data-testid="avgLoser"
                >
                  0
                </div>
              </div>
              <div
                class="MuiBox-root MuiBox-root-16"
              >
                <div>
                  Win Percentage:
                </div>
                <div
                  data-testid="winPerc"
                >
                  0
                  %
                </div>
              </div>
            </div>
            <div
              class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-2"
            />
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="MuiGrid-root MuiGrid-item MuiGrid-grid-sm-12 MuiGrid-grid-md-6"
    >
      <div
        class="MuiPaper-root paper MuiPaper-elevation1 MuiPaper-rounded"
      >
        <div
          class="MuiGrid-root MuiGrid-container"
        >
          <div
            class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-4"
          >
            <div
              class="MuiBox-root MuiBox-root-9"
            >
              <div>
                Trades:
              </div>
              <div
                data-testid="tradesCount"
              >
                1
              </div>
            </div>
            <div
              class="MuiBox-root MuiBox-root-10"
            >
              <div>
                Winning Trades:
              </div>
              <div
                data-testid="winningTradesCount"
              >
                0
              </div>
            </div>
            <div
              class="MuiBox-root MuiBox-root-11"
            >
              <div>
                Losing Trades:
              </div>
              <div
                data-testid="losingTradesCount"
              >
                0
              </div>
            </div>
            <div
              class="MuiBox-root MuiBox-root-12"
            >
              <div>
                Neutral Trades:
              </div>
              <div
                data-testid="neutralTradesCount"
              >
                1
              </div>
            </div>
          </div>
          <div
            class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-2"
          />
          <div
            class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-4"
          >
            <div
              class="MuiBox-root MuiBox-root-13"
            >
              <div>
                Profit:
              </div>
              <div
                class="green"
                data-testid="profit"
              >
                1
              </div>
            </div>
            <div
              class="MuiBox-root MuiBox-root-14"
            >
              <div>
                Average Winner:
              </div>
              <div
                class=""
                data-testid="avgWinner"
              >
                0
              </div>
            </div>
            <div
              class="MuiBox-root MuiBox-root-15"
            >
              <div>
                Average Loser:
              </div>
              <div
                class=""
                data-testid="avgLoser"
              >
                0
              </div>
            </div>
            <div
              class="MuiBox-root MuiBox-root-16"
            >
              <div>
                Win Percentage:
              </div>
              <div
                data-testid="winPerc"
              >
                0
                %
              </div>
            </div>
          </div>
          <div
            class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-2"
          />
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
